engine/crates/engine-v2/src/operation/logical_planner/selection_set.rs

```diff
@@ -553,10 +606,153 @@ impl<'schema, 'a> SelectionSetLogicalPlanner<'schema, 'a> {
+            }
+            _ => (),
         }
+
         Ok(())
     }
 
+    /// Replans the fields within a selection set by resetting their current plans and adding
+    /// new fields to be planned. This method may be invoked when the structure of the selection
+    /// set changes or when new fields become available for planning.
+    fn replan_fields(
+        &mut self,
+        selection_set_id: SelectionSetId,
+        plan_changes: Vec<(FieldId, Vec<FieldDefinition<'schema>>)>,
+        unplanned_fields: &mut HashMap<FieldId, FieldDefinition<'schema>>,
+    ) {
+        for (field_id, add_to_plan) in plan_changes {
+            unplanned_fields.remove(&field_id);
+            self.reset_field_plans(field_id);
+
+            let position = self.operation[selection_set_id]
+                .field_ids_ordered_by_parent_entity_id_then_position
+                .iter()
+                .position(|p| p == &field_id)
+                .unwrap();
+
+            self.operation[selection_set_id]
+                .field_ids_ordered_by_parent_entity_id_then_position
+                .remove(position);
+
+            for field_definition in add_to_plan {
+                let field = {
+                    let operation_field = self.walker().walk(field_id).as_ref().clone();
+                    let selection_set_id = operation_field.selection_set_id();
+
+                    ExtraField {
+                        edge: operation_field.response_edge(),
+                        definition_id: field_definition.id(),
+                        selection_set_id: selection_set_id.map(|id| self.clone_selection_set(id)),
+                        argument_ids: operation_field.argument_ids(),
+                        petitioner_location: operation_field.location(),
+                        parent_selection_set_id: operation_field.parent_selection_set_id(),
+                    }
+                };
+
+                let parent_selection_set_id = field.parent_selection_set_id;
+                let definition_id = field.definition_id;
+
+                self.operation.fields.push(Field::Extra(field));
+                let id = (self.operation.fields.len() - 1).into();
+                unplanned_fields.insert(id, field_definition);
+
+                self.field_to_logical_plan_id.push(None);
+                self.field_to_solved_requirement.push(None);
+
+                self.insert_field_in_parent_selection_set(parent_selection_set_id, definition_id, id);
+            }
+        }
+    }
+
+    /// Resets the planning state for a given field identified by `field_id`.
+    ///
+    /// This function updates the logical plan ID and resolved requirements for the specified
+    /// field, allowing it to be re-planned in subsequent operations. If the field has a selection
+    /// set, this process is recursive, resetting all fields within that selection set.
+    fn reset_field_plans(&mut self, field_id: FieldId) {
```

Comment:
would have used `mark_field_as_never_planned()` or something like this. Not a "reset" imho